### ISSUE TITLE
Last post tool use hook not triggered when maxTurns limit is reached

### ISSUE THREAD

**Issue by paradite** [anthropics/claude-agent-sdk-typescript#58]
there is a bug with post tool use hook:

If the agent reaches `maxTurns` set in the options, the last tool call's post call hook will not be triggered, instead, the agent will just send a `result` message with `error_max_turns`.

I have a minimal reproduction code here:

https://github.com/paradite/claude-agent-sdk-tool-hook-bug

If you run `npm install` and `npm start`, you can see that there are 2 tool call executions, but only one post tool call hook called:

```
number of tool call executions 2
number of tool call hook calls 1
```

## Versions

The bug is there in Claude Code version `1.0.120` (could be there earlier, but that's the version I was using).

I've tested latest version of agent sdk `@anthropic-ai/claude-agent-sdk": "0.1.30` this bug is still there.

## This is not due to input stream being closed

I also verified that this is not due to input stream being closed.

If we remove the max turn limit `maxTurns: 3`, the last post tool use hook is triggered correctly:

```
number of tool call executions 3
number of tool call hook calls 3
```

**Comment by qing-ant** [comment]
Verified fixed via live e2e on `@anthropic-ai/claude-agent-sdk@0.2.92` (CLI 2.1.92).

**Test:** `maxTurns: 1`, prompt that triggers two `Bash` tool calls in the single turn, with a `PostToolUse` hook matching `Bash`:

```js
import { query } from "@anthropic-ai/claude-agent-sdk";
let postToolUseCalls = 0, toolUseCount = 0;
const response = query({
  prompt: "Run `echo step1` then `echo step2` as two separate Bash tool calls.",
  options: {
    permissionMode: "bypassPermissions",
    maxTurns: 1,
    allowedTools: ["Bash"],
    hooks: {
      PostToolUse: [{ matcher: "Bash", hooks: [async () => { postToolUseCalls++; return { continue: true }; }] }],
    },
  },
});
for await (const m of response) {
  if (m.type === "assistant") for (const b of m.message.content) if (b.type === "tool_use") toolUseCount++;
}
```

**Result:**
```
[msg] tool_use #1: Bash
[msg] tool_use #2: Bash
[hook] PostToolUse #1: Bash
[hook] PostToolUse #2: Bash
[msg] result: error_max_turns
{"toolUseCount":2,"postToolUseCalls":2,"resultSubtype":"error_max_turns","pass":true}
```

PostToolUse fires for every tool execution before the `error_max_turns` result is emitted — the count now matches. In the current query loop, tool execution (where PostToolUse fires) completes before the maxTurns check, so the last hook is no longer dropped.

Closing as resolved. If you still see a mismatch on 0.2.92+ please reopen with a repro.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by brianantonelli** [comment]
We ran into this in production and ended up with a pattern that sidesteps the hook reliability issue entirely: give the 
agent a dedicated MCP submit tool and instruct it in the system prompt to call it proactively before running out of turns. 
 
The key insight is that if the agent calls the submit tool within its own turn, it doesn't matter whether the 
Stop/PostToolUse hook fires at session end - the submission has already happened through the normal tool-call path, which is reliable. The hook becomes a nice-to-have rather than a correctness requirement. 
 
Combined with detecting `error_max_turns` on the `ResultMessage` and doing a session-resume with `maxTurns=3` and a force-submit prompt as a backstop, this has been our most robust approach. The result field is absent on `error_max_turns` but `session_id`  is present, so the resume works.
